nat-lab/tests/test_vpn.py

```diff
@@ -219,3 +223,143 @@ async def test_vpn_reconnect(
+                    vpn_2_limits=ConnectionLimits(1, 1),
+                ),
+            )
+        )
+
+        client = await exit_stack.enter_async_context(
+            Client(
+                connection,
+                alpha,
+                adapter_type,
+                telio_features=TelioFeatures(boringtun_reset_connections=True),
+            ).run()
+        )
+
+        async def connect(
+            wg_server: dict,
+            connection_key: str,
+        ):
+            await testing.wait_long(
+                client.connect_to_vpn(
+                    wg_server["ipv4"], wg_server["port"], wg_server["public_key"]
+                )
+            )
+
+            await testing.wait_lengthy(
+                client.wait_for_state_peer(
+                    wg_server["public_key"], [State.Connected], [PathType.Direct]
+                )
+            )
+
+            await testing.wait_long(conn_tracker.wait_for_event(connection_key))
+
+            async with Ping(connection, serv_ip).run() as ping:
+                await testing.wait_long(ping.wait_for_next_ping())
+
+        await connect(
+            config.WG_SERVER,
+            "vpn_1",
+        )
+
+        output_notifier = OutputNotifier()
+
+        async def on_stdout(stdout: str) -> None:
+            for line in stdout.splitlines():
+                output_notifier.handle_output(line)
+
+        async def conntrack_on_stdout(stdout: str) -> None:
+            for line in stdout.splitlines():
+                if f"dst={serv_ip}" in line:
+                    output_notifier.handle_output(line)
+
+        sender_start_event = asyncio.Event()
+        close_wait_event = asyncio.Event()
+
+        output_notifier.notify_output(
+            "80 port [tcp/*] succeeded!",
+            sender_start_event,
+        )
+        output_notifier.notify_output("CLOSE", close_wait_event)
+
+        async with connection.create_process(
+            [
+                "conntrack",
+                "--family",
+                "ipv4" if ip_proto == IPProto.IPv4 else "ipv6",
+                "-E",
+            ]
+        ).run(stdout_callback=conntrack_on_stdout) as conntrack_proc:
+            await testing.wait_normal(conntrack_proc.wait_stdin_ready())
+
+            await exit_stack.enter_async_context(
+                connection.create_process(
+                    [
+                        "nc",
+                        "-nv",
+                        "-4" if ip_proto == IPProto.IPv4 else "-6",
+                        "-s",
+                        testing.unpack_optional(alpha.get_ip_address(ip_proto)),
+                        serv_ip,
+                        str(80),
+                    ]
+                ).run(stdout_callback=on_stdout, stderr_callback=on_stdout)
+            )
+            await testing.wait_normal(sender_start_event.wait())
+
+            await testing.wait_long(
+                client.disconnect_from_vpn(
```

Comment:
AFAIK, you should not need to disconnect from the existing VPN client, just issue new connect call.

Reply:
Actually python test framework fails without disconnecting first because it sets up interfaces